packages/style-engine/class-wp-style-engine.php

```diff
@@ -448,26 +474,72 @@ public function get_block_supports_styles( $block_styles, $options ) {
 		return $styles_output;
 	}
 
+	/**
+	 * Returns a stylesheet of CSS rules from a theme.json/global styles object.
+	 *
+	 * @param array $global_styles Styles object from theme.json.
+	 * @param array $options       array(
+	 *     'selector' => (string) When a selector is passed, the style engine will return a full CSS rule `$selector { ...rules }`, otherwise a concatenated string of properties and values.
+	 * );.
+	 *
+	 * @return string A stylesheet.
+	 */
+	public function generate_global_styles( $global_styles, $options ) {
+		if ( empty( $global_styles ) || ! is_array( $global_styles ) ) {
+			return null;
+		}
+
+		// The return stylesheet.
+		$global_stylesheet = '';
+
+		// Layer 0: Root.
+		$root_level_options_defaults = array(
+			'selector' => 'body',
+		);
+		$root_level_options          = wp_parse_args( $options, $root_level_options_defaults );
+		$root_level_styles           = $this->get_block_supports_styles(
+			$global_styles,
+			$root_level_options
+		);
+
+		if ( ! empty( $root_level_styles['css'] ) ) {
+			$global_stylesheet .= $root_level_styles['css'] . ' ';
+		}
+
+		// @TODO Layer 1: Elements.
```

Comment:
This might be a question for a follow-up PR instead, but just curious about the potential layers in follow-ups. Currently the theme JSON class iterates over nodes, so `generate_global_styles` is currently called once in this PR for the root. In a follow-up, would we call `generate_global_styles` in each loop as it happens now, or would it the iteration happen in the style engine?

(I wasn't sure which would be ideal, so curious to hear your thoughts about the role/responsibility of the theme JSON class versus style engine here).

One way I was imagining, incorporating the ideas from https://github.com/WordPress/gutenberg/pull/42222 is that (long term) we could potentially call the style engine once per iteration of the loop in the theme JSON class, but instead of it returning a css string immediately, it adds the rules to the style engine CSS rules store, and then once it finishes looping over each of the nodes, it calls the style engine method to retrieve all the generated CSS for output? I don't have any strong ideas about how any of that should work, so just thinking out loud 🙂

Reply:
Thanks for these thoughts @andrewserong !

> In a follow-up, would we call generate_global_styles in each loop as it happens now, or would it the iteration happen in the style engine?

Very good questions.

To be honest, I don't know. 😄 

The main motivation behind this PR was to experiment with throwing the entire merged theme.json at the style engine and generating an entire "global styles" _stylesheet_. I think this could be a cool feature of the public API anyway.

So you're right, this approach won't work in the current `foreach ( $block_nodes as $metadata )` loop, but we eventually wouldn't need those loops if the style engine could handle an entire theme.json.

> we could potentially call the style engine once per iteration of the loop in the theme JSON class, but instead of it returning a css string immediately, it adds the rules to the style engine CSS rules store, and then once it finishes looping over each of the nodes, it calls the style engine method to retrieve all the generated CSS for output

This sounds like it would be a more sensible approach, especially if it keeps any custom values parsing in WP_Theme_JSON for now.

It's also similar to the approach we had in mind for block supports, so it would make doubly more sense to have a global styles analogue.

We could put this PR to the side and concentrate our efforts on the store, which we'll need first anyway. What do you think?

Reply:
> We could put this PR to the side and concentrate our efforts on the store, which we'll need first anyway. What do you think?

Yeah, that might be a good next step. Each of these explorations is a great way of working out which puzzle piece to do next, and I think the store is probably the thing that's most helpful for illuminating the path forward for this PR. It'll probably also be the thing that highlights the value of switching over to the style engine? (The clearer separation between generating and outputting styles)

But this PR's great for highlighting how we can hook in the style engine, and it's really exciting seeing all the style engine efforts starting to come together like this!